runtime/autoload/man.vim

```diff
@@ -11,127 +16,217 @@ catch /E145:/
     silent keepjumps 1delete _
   endwhile
   while getline('$') =~# '^\s*$'
     silent keepjumps $delete _
   endwhile
-  setlocal nomodified
+  keepjumps 1
   setlocal filetype=man
+endfunction
 
-  if invoked_from_man || editcmd ==# 'tabnew'
-    call s:set_window_local_options()
+function s:cmd_args(sect, page) abort
+  if !empty(a:sect)
+    return s:man_sect_arg.' '.a:sect.' '.a:page
   endif
+  return a:page
 endfunction
 
-function s:set_window_local_options() abort
-  setlocal colorcolumn=0 foldcolumn=0 nonumber
-  setlocal nolist norelativenumber nofoldenable
+function! s:error(msg) abort
+  redrawstatus!
+  echon "man: "
+  echohl ErrorMsg
+  echon a:msg
+  echohl None
 endfunction
 
-function man#pop_page() abort
-  if s:man_tag_depth > 0
-    let s:man_tag_depth = s:man_tag_depth - 1
-    exec "let s:man_tag_buf=s:man_tag_buf_".s:man_tag_depth
-    exec "let s:man_tag_lin=s:man_tag_lin_".s:man_tag_depth
-    exec "let s:man_tag_col=s:man_tag_col_".s:man_tag_depth
-    exec s:man_tag_buf."b"
-    exec s:man_tag_lin
-    exec "norm! ".s:man_tag_col."|"
-    exec "unlet s:man_tag_buf_".s:man_tag_depth
-    exec "unlet s:man_tag_lin_".s:man_tag_depth
-    exec "unlet s:man_tag_col_".s:man_tag_depth
-    unlet s:man_tag_buf s:man_tag_lin s:man_tag_col
+function! man#Complete(ArgLead, CmdLine, CursorPos) abort
+  let args = split(a:CmdLine)
+  let l = len(args)
+  " if the cursor (|) is at ':Man printf(|' then
+  " make sure to display the section. See s:get_candidates
+  let fpage = 0
+  " if already completed a manpage, we return
+  if (l > 1 && args[1] =~# ')\f*$') || l > 3 || a:ArgLead =~# ')$'
+    return
+  elseif l == 3
+    " cursor (|) is at ':Man 3 printf |'
+    if empty(a:ArgLead)
+      return
+    endif
+    let sect = args[1]
+    let page = args[2]
+  elseif l == 2
+    " cursor (|) is at ':Man 3 |'
+    if empty(a:ArgLead)
+      let page = ""
+      let sect = args[1]
+    elseif a:ArgLead =~# '^\f\+(\f*$'
+      " cursor (|) is at ':Man printf(|'
+      let tmp = split(a:ArgLead, '(')
+      let page = tmp[0]
+      let sect = substitute(get(tmp, 1, '*'), ')$', '', '').'*'
+      let fpage = 1
+    else
+      " cursor (|) is at ':Man printf
+      let page = args[1]
+      let sect = '*'
+    endif
+  else
+    let page = ''
+    let sect = '*'
   endif
+  return s:get_candidates(page, sect, fpage)
 endfunction
 
-" Expects a string like 'access' or 'access(2)'.
-function s:parse_page_and_section(sect, str) abort
-  try
-    let [page, sect] = matchlist(a:str, '\v\C([-.[:alnum:]_]+)%(\(([-.[:alnum:]_]+)\))?')[1:2]
-    if empty(sect)
-      let sect = a:sect
+function! s:get_candidates(page, sect, fpage) abort
+  let mandirs = s:MANDIRS()
+  let candidates = globpath(mandirs, "*/" . a:page . "*." . a:sect, 0, 1)
+  let find = '\(.\+\)\.\%('.s:man_extensions.'\)\@!\'
+  " if the page is a path, complete files
+  if a:sect ==# '*' && a:page =~# '\/'
+    "TODO why does this complete the last one automatically
```

Comment:
@justinmk @noahfrederick can you guys test this? Try `:Man ./<TAB>`. For some reason, it completes the last result automatically for me, yet this doesn't happen with `:Man <TAB>`?